libs/libcommon/include/common/fixed_mem_eq.h

```diff
@@ -0,0 +1,125 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <common/defines.h>
+
+#include <cstddef>
+#include <cstdint>
+
+namespace mem_utils
+{
+
+FLATTEN_INLINE_PURE
+constexpr inline bool memcmp_eq0(const char *, const char *)
+{
+    return true;
+}
+
+FLATTEN_INLINE_PURE
+inline bool memcmp_eq1(const char * a, const char * b)
+{
+    return a[0] == b[0];
+}
+
+FLATTEN_INLINE_PURE
+inline bool memcmp_eq2(const char * a, const char * b)
+{
+    const uint16_t ax = *reinterpret_cast<const uint16_t *>(a);
+    const uint16_t bx = *reinterpret_cast<const uint16_t *>(b);
+    return ax == bx;
+}
+
+FLATTEN_INLINE_PURE
+inline bool memcmp_eq3(const char * a, const char * b)
+{
+    return memcmp_eq2(a, b) & memcmp_eq2(a + 1, b + 1);
+}
+
+FLATTEN_INLINE_PURE
+inline bool memcmp_eq4(const char * a, const char * b)
+{
+    const uint32_t ax = *reinterpret_cast<const uint32_t *>(a);
+    const uint32_t bx = *reinterpret_cast<const uint32_t *>(b);
+    return ax == bx;
+}
+
+FLATTEN_INLINE_PURE
+inline bool memcmp_eq8(const char * a, const char * b)
+{
+    const uint64_t ax = *reinterpret_cast<const uint64_t *>(a);
+    const uint64_t bx = *reinterpret_cast<const uint64_t *>(b);
+    return ax == bx;
+}
+
+template <size_t k>
+ALWAYS_INLINE inline bool memcmp_eq_fixed_size(const char * a, const char * b)
+{
+#ifdef M
+    static_assert(false, "`M` is defined");
+#else
+#define M(s)                       \
+    else if constexpr (k == (s))   \
+    {                              \
+        return memcmp_eq##s(a, b); \
+    }
+#endif
+
+    static_assert(k >= 0);
+    static_assert(k <= 32);
+
+    if constexpr (k > 16)
+    {
+        return memcmp_eq_fixed_size<16>(a, b) & memcmp_eq_fixed_size<k - 16>(a + 16, b + 16);
```

Comment:
Any consideration about using bitwise and (`&`) instead of logical and (`&&`)?

Reply:
Just saw the other comment address this question.

But maybe we can add some comments here.

Reply:
Same as https://github.com/pingcap/tiflash/pull/5569#discussion_r945431409